File: docs/algorithms/code/algorithms/largestValues.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function largestValues(root: TreeNode | null): number[] {
  if (!root) {
    return [];
  }

  const stack: TreeNode[] = [root];
  const res: number[] = [];

  while (stack.length) {
    let size = stack.length;
    let max = -Infinity;

    while (size--) {
      const node = stack.shift();
      max = node.val > max ? node.val : max;
      if (node.left) {
        stack.push(node.left);
      }

      if (node.right) {
        stack.push(node.right);
      }
    }

    res.push(max);
  }

  return res;
};

File: docs/algorithms/code/algorithms/largestValues.go
// 思路：层序遍历，每一层获取中位数再塞进去
// 复杂度分析：时间复杂度 O(n), 空间复杂度 O(n)
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func largestValues(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}

	res := []int{};
	queue := []*TreeNode{root};
	for len(queue) > 0 {
		size := len(queue);
		max := queue[0].Val;
		for i := 0; i < size; i++ {
			node := queue[0];
			queue = queue[1:];
			if node.Val > max {
				max = node.Val;
			}
			if node.Left != nil {
				queue = append(queue, node.Left);
			}
			if node.Right != nil {
				queue = append(queue, node.Right);
			}
		}
		res = append(res, max);
	}
	return res;
}